README
This is the README file for the NetStore package.

This software comes with ABSOLUTELY NO WARRANTY. See the file COPYING for
details.


NEWS
The NEWS file for the NetStore package.

-------------------------------------------------------------------------------


src/tests/manual/datetime.h
// datetime.h
//
// Parse and write dates/times in various standard formats. 
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef DATETIME_H
#define DATETIME_H

#include <qstring.h>
#include <qdatetime.h>

//
// Function Prototypes
//

//
// Auto-detect the format (XML xs:dateTime or RFC822)
//
QDateTime ParseDateTime(const QString &str,bool *ok);

//
// XML xs:date format
//
QDate ParseXmlDate(const QString &str,bool *ok);
QString WriteXmlDate(const QDate &date);

//
// XML xs:time format
//
QTime ParseXmlTime(const QString &str,bool *ok,int *day_offset=NULL);
QString WriteXmlTime(const QTime &time);

//
// XML xs:dateTime format
//
QDateTime ParseXmlDateTime(const QString &str,bool *ok);
QString WriteXmlDateTime(const QDateTime &dt);

//
// RFC822 date/time format
//
QDateTime ParseRfc822DateTime(const QString &str,bool *ok);
QString WriteRfc822DateTime(const QDateTime &dt);

//
// Returns the UTC offset of the curently configured timezone (seconds)
//
int TimeZoneOffset();


#endif  // DATETIME_H


src/tests/manual/cmdswitch.h
// cmdswitch.h
//
// Process Command-Line Switches
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef CMDSWITCH_H
#define CMDSWITCH_H

#include <vector>
#include <qstring.h>

class CmdSwitch
{
 public:
  CmdSwitch(int argc,char *argv[],const char *modname,const char *usage);
  unsigned keys() const;
  QString key(unsigned n) const;
  QString value(unsigned n) const;
  bool processed(unsigned n) const;
  void setProcessed(unsigned n,bool state);
  bool allProcessed() const;
  bool debugActive() const;

 private:
  std::vector<QString> switch_keys;
  std::vector<QString> switch_values;
  std::vector<bool> switch_processed;
  bool switch_debug;
};


#endif  // CMDSWITCH_H


src/tests/manual/datetime.cpp
// datetime.cpp
//
// Parse and write dates/times in various standard formats. 
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <stdio.h>
#include <time.h>

#include <QMap>
#include <QStringList>

#include "datetime.h"

QString __rddatetime_month_names[]=
  {"Jan","Feb","Mar","Apr","Mar","Jun","Jul","Aug","Sep","Oct","Nov","Dec"};
QString __rddatetime_dow_names[]=
  {"Mod","Tue","Wed","Thu","Fri","Sat","Sun"};

//
// Auto-detect the format (XML xs:dateTime or RFC822)
//
QDateTime ParseDateTime(const QString &str,bool *ok)
{
  if(str.trimmed().contains(" ")) {
    return ParseRfc822DateTime(str,ok);
  }
  return ParseXmlDateTime(str,ok);
}


//
// XML xs:date format
//
QDate ParseXmlDate(const QString &str,bool *ok)
{
  QDate ret=QDate::fromString(str,"yyyy-MM-dd");
  if(ok!=NULL) {
    *ok=ret.isValid();
  }
  return ret;
}


QString WriteXmlDate(const QDate &date)
{
  return date.toString("yyyy-MM-dd");
}


//
// XML xs:time format
//
QTime ParseXmlTime(const QString &str,bool *ok,int *day_offset)
{
  QTime ret;
  QStringList f0;
  QStringList f1;
  QStringList f2;
  int tz=0;
  QTime time;
  QTime tztime;

  if(ok!=NULL) {
    *ok=false;
  }
  if(day_offset!=NULL) {
    *day_offset=0;
  }
  f0=str.trimmed().split(" ");
  if(f0.size()!=1) {
    if(ok!=NULL) {
      *ok=false;
    }
    return ret;
  }

  if(f0[0].right(1).toLower()=="z") {  // GMT
    tz=-TimeZoneOffset();
    f0[0]=f0[0].left(f0[0].length()-1);
    f2=f0[0].split(":");
  }
  else {
    f1=f0[0].split("+");
    if(f1.size()==2) {   // GMT+
      f2=f1[1].split(":");
      if(f2.size()==2) {
	tztime=QTime(f2[0].toInt(),f2[1].toInt(),0);
	if(tztime.isValid()) {
	  tz=-TimeZoneOffset()-QTime(0,0,0).secsTo(tztime);
	}
      }
      else {
	if(ok!=NULL) {
	  *ok=false;
	}
	return QTime();
      }
    }
    else {
      f1=f0[0].split("-");
      if(f1.size()==2) {   // GMT-
	f2=f1[1].split(":");
	if(f2.size()==2) {
	  tztime=QTime(f2[0].toInt(),f2[1].toInt(),0);
	  if(tztime.isValid()) {
	    tz=-TimeZoneOffset()+QTime(0,0,0).secsTo(tztime);
	  }
	}
	else {
	  if(ok!=NULL) {
	    *ok=false;
	  }
	  return QTime();
	}
      }
    }
    f2=f1[0].split(":");
  }
  if(f2.size()==3) {
    QStringList f3=f2[2].split(".");
    time=QTime(f2[0].toInt(),f2[1].toInt(),f2[2].toInt());
    if(time.isValid()) {
      ret=time.addSecs(tz);
      if(day_offset!=NULL) {
	if((tz<0)&&((3600*time.hour()+60*time.minute()+time.second())<(-tz))) {
	  *day_offset=-1;
	}
	if((tz>0)&&(86400-((3600*time.hour()+60*time.minute()+time.second()))<tz)) {
	  *day_offset=1;
	}
      }
      if(ok!=NULL) {
	*ok=true;
      }
    }
  }
  return ret;
}


QString WriteXmlTime(const QTime &time)
{
  int utc_off=TimeZoneOffset();
  QString tz_str="-";
  if(utc_off<0) {
    tz_str="+";
  }
  tz_str+=QString().
    sprintf("%02d:%02d",utc_off/3600,(utc_off-3600*(utc_off/3600))/60);

  return time.toString("hh:mm:ss")+tz_str;
}


//
// XML xs:dateTime format
//
QDateTime ParseXmlDateTime(const QString &str,bool *ok)
{
  QDateTime ret;
  QStringList list;
  QStringList f0;
  QStringList f1;
  QStringList f2;
  int day;
  int month;
  int year;
  QTime time;
  bool lok=false;
  int day_offset=0;

  if(ok!=NULL) {
    *ok=false;
  }

  f0=str.trimmed().split(" ");
  if(f0.size()!=1) {
    if(ok!=NULL) {
      *ok=false;
    }
    return ret;
  }
  f1=f0[0].split("T");
  if(f1.size()<=2) {
    f2=f1[0].split("-");
    if(f2.size()==3) {
      year=f2[0].toInt(&lok);
      if(lok&&(year>0)) {
	month=f2[1].toInt(&lok);
	if(lok&&(month>=1)&&(month<=12)) {
	  day=f2[2].toInt(&lok);
	  if(lok&&(day>=1)&&(day<=31)) {
	    if(f1.size()==2) {
	      time=ParseXmlTime(f1[1],&lok,&day_offset);
	      if(lok) {
		ret=QDateTime(QDate(year,month,day),time).addDays(day_offset);
		if(ok!=NULL) {
		  *ok=true;
		}
	      }
	    }
	  }
	}
      }
    }
  }
  return ret;
}


QString WriteXmlDateTime(const QDateTime &dt)
{
  return WriteXmlDate(dt.date())+"T"+WriteXmlTime(dt.time());
}


//
// RFC822 date/time format
//
QDateTime ParseRfc822DateTime(const QString &str,bool *ok)
{
  QStringList f0=str.trimmed().split(" ",QString::SkipEmptyParts);

  //
  // Remove useless day-of-the-week tag
  //
  if(f0.size()==6) {
    f0.removeFirst();
  }
  if(f0.size()!=5) {
    if(ok!=NULL) {
      *ok=false;
    }
    return QDateTime();
  }

  //
  // Read Date
  //
  int month=-1;
  for(int i=0;i<7;i++) {
    if(f0.at(1).toLower()==__rddatetime_month_names[i].toLower()) {
      month=i;
    }
  }
  if(month<0) {
    if(ok!=NULL) {
      *ok=false;
    }
    return QDateTime();
  }
  if(f0.at(2).length()==2) {
    f0[2]="19"+f0.at(2);
  }
  QDate date(f0.at(2).toInt(),month+1,f0.at(0).toInt());
  if(!date.isValid()) {
    if(ok!=NULL) {
      *ok=false;
    }
    return QDateTime();
  }

  //
  // Read Time
  //
  QTime time=QTime::fromString(f0.at(3),"hh:mm:ss");
  if(!time.isValid()) {
    if(ok!=NULL) {
      *ok=false;
    }
    return QDateTime();
  }
  
  //
  // Read TZ Offset
  //
  bool ok1=false;
  bool ok2=false;
  int tz_offset=3600*f0.at(4).mid(1,2).toInt(&ok1)+
    60*f0.at(4).right(2).toInt(&ok2);
  if((f0.at(4).length()==5)&&(ok1)&&(ok2)) {
    if(f0.at(4).left(1)=="+") {
      tz_offset=-tz_offset;
    }
    else {
      if(f0.at(4).left(1)!="-") {
	if(ok!=NULL) {
	  *ok=false;
	}
	return QDateTime();
      }
    }
  }
  else {
    //
    // Try legacy timezone names (see RFC822 section 5.1)
    //
    QMap<QString,int> zones;
    zones["ut"]=0;
    zones["gmt"]=0;
    zones["z"]=0;
    zones["est"]=-5;
    zones["edt"]=-4;
    zones["cst"]=-6;
    zones["cdt"]=-5;
    zones["mst"]=-7;
    zones["mdt"]=-6;
    zones["pst"]=-8;
    zones["pdt"]=-7;
    // "Military" zone IDs are not implemented (see RFC1123 section 5.2.14)
    if(zones.value(f0.at(4).toLower(),1)>0) {
      if(ok!=NULL) {
	*ok=false;
      }
      return QDateTime();
    }
    tz_offset=-3600*zones.value(f0.at(4).toLower());
  }
  
  if(ok!=NULL) {
    *ok=true;
  }

  return QDateTime(date,time).addSecs(tz_offset-TimeZoneOffset());
}


QString WriteRfc822DateTime(const QDateTime &dt)
{
  int utc_off=TimeZoneOffset();
  QString tz_str="-";
  if(utc_off<0) {
    tz_str="+";
  }
  tz_str+=QString().
    sprintf("%02d%02d",utc_off/3600,(utc_off-3600*(utc_off/3600))/60);

  return __rddatetime_dow_names[dt.date().dayOfWeek()-1]+", "+
    QString().sprintf("%d ",dt.date().day())+
    __rddatetime_month_names[dt.date().month()-1]+" "+
    QString().sprintf("%04d ",dt.date().year())+
    dt.toString("hh:mm:ss")+" "+
    tz_str;
}


//
// Returns the UTC offset of the curently configured timezone (seconds)
//
int TimeZoneOffset()
{
  time_t t=time(&t);
  struct tm *tm=localtime(&t);
  time_t local_time=3600*tm->tm_hour+60*tm->tm_min+tm->tm_sec;
  tm=gmtime(&t);
  time_t gmt_time=3600*tm->tm_hour+60*tm->tm_min+tm->tm_sec;

  int offset=gmt_time-local_time;
  if(offset>43200) {
    offset=offset-86400;
  }
  if(offset<-43200) {
    offset=offset+86400;
  }

  return offset;
}




autogen.sh
#!/bin/sh
##
##    (C) Copyright 2002-2020 Fred Gleason <fredg@paravelsystems.com>
##
##    Adapted from './autogen.sh' in the Jack Audio Connection Kit.
##    Copyright (C) 2001-2003 Paul Davis, et al.
##
##    This program is free software; you can redistribute it and/or modify
##    it under the terms of version 2 of the GNU General Public License as
##    published by the Free Software Foundation;
##
##    This program is distributed in the hope that it will be useful,
##    but WITHOUT ANY WARRANTY; without even the implied warranty of
##    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##    GNU General Public License for more details.
##
##    You should have received a copy of the GNU General Public License
##    along with this program; if not, write to the Free Software
##    Foundation, Inc., 59 Temple Place, Suite 330, 
##    Boston, MA  02111-1307  USA
##

mkdir -p m4

libtoolize=libtoolize
if which glibtoolize > /dev/null 2>&1; then
      libtoolize=glibtoolize
fi

$libtoolize --force 2>&1 | sed '/^You should/d' || {
    echo "libtool failed, exiting..."
    exit 1
}

aclocal $ACLOCAL_FLAGS || {
    echo "aclocal \$ACLOCAL_FLAGS where \$ACLOCAL_FLAGS= failed, exiting..."
    exit 1
}

automake --add-missing -Wno-portability || {
    echo "automake --add-missing failed, exiting..."
    exit 1
}

autoconf || {
    echo "autoconf failed, exiting..."
    exit 1
}


src/ns/nsnetstore.h
// nsnetstore.h
//
// Client support for Akamai NetStore 4
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef NSNETSTORE_H
#define NSNETSTORE_H

#include <curl/curl.h>

#include <QByteArray>
#include <QDateTime>
#include <QList>
#include <QObject>
#include <QStringList>

#include <nsnetstoreobject.h>

#define NSNETSTORE_AUTH_ENCRYPTION_FORMAT 5

class NSNetStore : public QObject
{
 Q_OBJECT
 public:
 enum Action {Delete=0,Dir=1,Download=2,Du=3,List=4,Mkdir=5,Mtime=6,
	      QuickDelete=7,Rename=8,Rmdir=9,Stat=10,Symlink=11,Upload=12,
	      LastAction=13};
  NSNetStore(const QString &hostname,const QString &cp_code,
	     const QString &keyname,const QString &secret,QObject *parent=0);
  NSNetStore(QObject *parent=0);
  QString hostName() const;
  void setHostName(const QString &str);
  QString cpCode() const;
  void setCpCode(const QString &str);
  QString keyName() const;
  void setKeyName(const QString &str);
  QString secret() const;
  void setSecret(const QString &str);
  bool dir(QList<NSNetStoreObject> &objs,const QString &path);
  static QString sigHeader1(const QDateTime &dt,unsigned unique_id,
			    const QString &keyname);
  static QString sigHeader2(const QString &hdr1,Action act,const QString &path,
			    const QString &secret,
			    const QStringList &hdrs=QStringList());
  static QString actionText(Action act);

 private:
  static QString GetLastSslErrorString();
  struct curl_slist *AddAuthHeaders(struct curl_slist *headers,
				    Action act,const QString &path,
				    const QStringList &hdrs=QStringList())
    const;
  QString c_host_name;
  QString c_cp_code;
  QString c_key_name;
  QString c_secret;
  friend size_t __NSNetStore_DownloadCB(void *ptr,size_t size,size_t nmemb,
				     void *userdata);
  QByteArray c_cb_data;
};


#endif  // NSNETSTORE_H


src/ns/nsnetstoreobject.h
// nsnetstoreobject.h
//
// Container class for Akamai ObjectStore (NS4) object attributes
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#ifndef NSNETSTOREOBJECT_H
#define NSNETSTOREOBJECT_H

#include <stdint.h>

#include <QDateTime>
#include <QString>

class NSNetStoreObject
{
 public:
  enum Type {File=0,Symlink=1,Dir=2,LastType=3};
  NSNetStoreObject();
  Type type() const;
  void setType(Type type);
  QString typeText() const;
  QString name() const;
  void setName(const QString &str);
  uint64_t size() const;
  void setSize(uint64_t bytes);
  QDateTime modifiedDateTime() const;
  void setModifiedDateTime(const QDateTime &dt);
  QString md5Sum() const;
  void setMd5Sum(const QString &hash);
  bool isImplicit() const;
  void setImplicit(bool state);
  bool isType(const NSNetStoreObject &other);
  static QString typeText(Type type);

 private:
  Type c_type;
  QString c_name;
  uint64_t c_size;
  QDateTime c_modified_datetime;
  QString c_md5_sum;
  bool c_implicit;
};


#endif  // NSNETSTOREOBJECT_H


src/ns/nsnetstoreobject.cpp
// nsnetstoreobject.cpp
//
// Container class for Akamai ObjectStore (NS4) object attributes
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include "nsnetstoreobject.h"

NSNetStoreObject::NSNetStoreObject()
{
  c_type=NSNetStoreObject::LastType;
  c_size=0;
  c_implicit=false;
}


NSNetStoreObject::Type NSNetStoreObject::type() const
{
  return c_type;
}


void NSNetStoreObject::setType(NSNetStoreObject::Type type)
{
  c_type=type;
}


QString NSNetStoreObject::typeText() const
{
  return NSNetStoreObject::typeText(c_type);
}


QString NSNetStoreObject::name() const
{
  return c_name;
}


void NSNetStoreObject::setName(const QString &str)
{
  c_name=str;
}


uint64_t NSNetStoreObject::size() const
{
  return c_size;
}


void NSNetStoreObject::setSize(uint64_t bytes)
{
  c_size=bytes;
}


QDateTime NSNetStoreObject::modifiedDateTime() const
{
  return c_modified_datetime;
}


void NSNetStoreObject::setModifiedDateTime(const QDateTime &dt)
{
  c_modified_datetime=dt;
}


QString NSNetStoreObject::md5Sum() const
{
  return c_md5_sum;
}


void NSNetStoreObject::setMd5Sum(const QString &hash)
{
  c_md5_sum=hash;
}


bool NSNetStoreObject::isImplicit() const
{
  return c_implicit;
}


void NSNetStoreObject::setImplicit(bool state)
{
  c_implicit=state;
}


bool NSNetStoreObject::isType(const NSNetStoreObject &other)
{
  return other.type()==c_type;
}


QString NSNetStoreObject::typeText(NSNetStoreObject::Type type)
{
  QString ret="unknown";

  switch(type) {
  case NSNetStoreObject::File:
    ret="file";
    break;

  case NSNetStoreObject::Symlink:
    ret="symlink";
    break;

  case NSNetStoreObject::Dir:
    ret="dir";
    break;

  case NSNetStoreObject::LastType:
    break;
  }

  return ret;
}


src/ns/nsnetstore.cpp
// nsnetstore.cpp
//
// Client support for Akamai NetStore 4
//
//   (C) Copyright 2020 Fred Gleason <fredg@paravelsystems.com>
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License version 2 as
//   published by the Free Software Foundation.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public
//   License along with this program; if not, write to the Free Software
//   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
//

#include <openssl/engine.h>
#include <openssl/err.h>
#include <openssl/hmac.h>
#include <openssl/evp.h>

#include "nsnetstore.h"

//
// CURL Callbacks
//
size_t __NSNetStore_DownloadCB(void *ptr,size_t size,size_t nmemb,void *userdata)
{
  QByteArray *data=(QByteArray *)userdata;
  *data+=QByteArray((const char *)ptr,size*nmemb);

  printf("DATA: %s\n",data->constData());

  return size*nmemb;
}

NSNetStore::NSNetStore(const QString &hostname,const QString &cp_code,
		       const QString &keyname,const QString &secret,
		       QObject *parent)
  : QObject(parent)
{
  c_host_name=hostname;
  c_cp_code=cp_code;
  c_key_name=keyname;
  c_secret=secret;
}


QString NSNetStore::hostName() const
{
  return c_host_name;
}


void NSNetStore::setHostName(const QString &str)
{
  c_host_name=str;
}


QString NSNetStore::cpCode() const
{
  return c_cp_code;
}


void NSNetStore::setCpCode(const QString &str)
{
  c_cp_code=str;
}


QString NSNetStore::keyName() const
{
  return c_key_name;
}


void NSNetStore::setKeyName(const QString &str)
{
  c_key_name=str;
}


QString NSNetStore::secret() const
{
  return c_secret;
}


void NSNetStore::setSecret(const QString &str)
{
  c_secret=str;
}


bool NSNetStore::dir(QList<NSNetStoreObject> &objs,const QString &path)
{
  QString url;
  long response_code;
  CURL *curl=NULL;
  CURLcode curl_err;
  struct curl_slist *cheaders=NULL;

  if((curl=curl_easy_init())==NULL) {
    return false;
  }
  c_cb_data=QByteArray();

  //
  // Add Custom Headers
  //
  cheaders=
    curl_slist_append(cheaders,
		      (QString("X-Akamai-ACS-Action: ")+
		       "version=1"+
		       "&action=dir"+
		       "&format=xml").toUtf8().constData());
  cheaders=AddAuthHeaders(cheaders,NSNetStore::Dir,path);

  url=QString("http://")+c_host_name+"-nsu.akamaihd.net/"+c_cp_code+path;
  curl_easy_setopt(curl,CURLOPT_URL,url.toUtf8().constData());
  printf("URL: %s\n",url.toUtf8().constData());
  curl_easy_setopt(curl,CURLOPT_HTTPHEADER,cheaders);
  curl_easy_setopt(curl,CURLOPT_WRITEFUNCTION,__NSNetStore_DownloadCB);
  curl_easy_setopt(curl,CURLOPT_WRITEDATA,&c_cb_data);
  curl_easy_setopt(curl,CURLOPT_FOLLOWLOCATION,1L);
  //  curl_easy_setopt(curl,CURLOPT_USERAGENT,
  //		   conv_config->userAgent().toUtf8().constData());
  //curl_easy_setopt(curl,CURLOPT_TIMEOUT,RD_CURL_TIMEOUT);

  if((curl_err=curl_easy_perform(curl))==CURLE_OK) {
    printf("DOWNLOADED: %s\n",c_cb_data.constData());
  }
  else {
    printf("CURL ERROR\n");
    curl_easy_cleanup(curl);
    return false;
  }

  curl_easy_getinfo(curl,CURLINFO_RESPONSE_CODE,&response_code);
  curl_easy_cleanup(curl);
  curl_slist_free_all(cheaders);

  printf("RESPONSE_CODE: %ld\n",response_code);

  switch(response_code) {
  case 200:
    break;

  case 404:
    return false;

  default:
    return false;
  }

  return true;
}


QString NSNetStore::sigHeader1(const QDateTime &dt,unsigned unique_id,
			       const QString &keyname)
{
  QString ret="";

  ret+=QString().sprintf("%u, ",NSNETSTORE_AUTH_ENCRYPTION_FORMAT);
  ret+="0.0.0.0, ";
  ret+="0.0.0.0, ";
  ret+=QString().sprintf("%u, ",dt.toTime_t());
  ret+=QString().sprintf("%u, ",unique_id);
  ret+=keyname;

  return ret;
}


QString NSNetStore::sigHeader2(const QString &hdr1,NSNetStore::Action act,
			       const QString &path,const QString &secret,
			       const QStringList &hdrs)
{
  QString ret="";
  QString raw="";
  HMAC_CTX ctx;
  unsigned char md_value[EVP_MAX_MD_SIZE];
  unsigned md_len=0;

  //
  // Build raw data
  //
  raw+=hdr1;
  raw+=path+"\n";
  raw+="x-akamai-acs-action:version=1&action="+NSNetStore::actionText(act);
  for(int i=0;i<hdrs.size();i++) {
    raw+="&"+hdrs.at(i);
  }
  raw+="\n";

  //
  // Genrate the HMAC-SHA256 hash
  //
  ENGINE_load_builtin_engines();
  ENGINE_register_all_complete();

  HMAC_CTX_init(&ctx);
  if(HMAC_Init_ex(&ctx,secret.toUtf8(),secret.toUtf8().size(),EVP_sha256(),
		  NULL)==0) {
    fprintf(stderr,
	    "NSNetStore::sigHeader2(): HMAC_Init_ex() returned an error [%s]\n",
	    GetLastSslErrorString().toUtf8().constData());
  }
  if(HMAC_Update(&ctx,(const unsigned char *)raw.toUtf8().constData(),
		 raw.toUtf8().size())==0) {
    fprintf(stderr,
	    "NSNetStore::sigHeader2(): HMAC_Update() returned an error [%s]\n",
	    GetLastSslErrorString().toUtf8().constData());
  }
  if(HMAC_Final(&ctx,md_value,&md_len)==0) {
    fprintf(stderr,
	    "NSNetStore::sigHeader2(): HMAC_Final() returned an error [%s]\n",
	    GetLastSslErrorString().toUtf8().constData());
  }
  HMAC_CTX_cleanup(&ctx);

  return QByteArray((const char *)md_value,md_len).toBase64();
}


QString NSNetStore::actionText(NSNetStore::Action act)
{
  switch(act) {
  case NSNetStore::Delete:
    return QString("delete");

  case NSNetStore::Dir:
    return QString("dir");

  case NSNetStore::Download:
    return QString("download");

  case NSNetStore::Du:
    return QString("du");

  case NSNetStore::List:
    return QString("list");

  case NSNetStore::Mkdir:
    return QString("mkdir");

  case NSNetStore::Mtime:
    return QString("mtime");

  case NSNetStore::QuickDelete:
    return QString("quick-delete");

  case NSNetStore::Rename:
    return QString("rename");

  case NSNetStore::Rmdir:
    return QString("rmdir");

  case NSNetStore::Stat:
    return QString("stat");

  case NSNetStore::Symlink:
    return QString("symlink");

  case NSNetStore::Upload:
    return QString("upload");

  case NSNetStore::LastAction:
    break;
  }

  return QString("unknown");
}


QString NSNetStore::GetLastSslErrorString()
{
  char err_str[120];

  ERR_error_string_n(ERR_peek_last_error(),err_str,120);
  return QString(err_str);
}


struct curl_slist *NSNetStore::AddAuthHeaders(struct curl_slist *cheaders,
					      NSNetStore::Action act,
					      const QString &path,
					      const QStringList &hdrs) const
{
  QString hdr1=
    NSNetStore::sigHeader1(QDateTime::currentDateTime(),random(),c_key_name);

  cheaders=curl_slist_append(cheaders,
			    (QString("X-Akamai-ACS-Auth-Data: ")+
			     hdr1).toUtf8().constData());
  cheaders=curl_slist_append(cheaders,
			    (QString("X-Akamai-ACS-Auth-Sign:")+
			     NSNetStore::sigHeader2(hdr1,act,path,c_secret,
						   hdrs)).toUtf8().constData());
  return cheaders;
}
